File: spigot/src/main/java/us/minevict/mvutil/spigot/MvPlugin.java
package us.minevict.mvutil.spigot;

import co.aikar.commands.BaseCommand;
import co.aikar.commands.PaperCommandManager;
import co.aikar.idb.Database;
import co.aikar.taskchain.BukkitTaskChainFactory;
import co.aikar.taskchain.TaskChain;
import co.aikar.taskchain.TaskChainFactory;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.TimeUnit;
import org.bukkit.event.HandlerList;
import org.bukkit.event.Listener;
import org.bukkit.plugin.java.JavaPlugin;
import org.bukkit.scheduler.BukkitTask;
import org.jetbrains.annotations.NotNull;
import us.minevict.mvutil.common.IMvPlugin;
import us.minevict.mvutil.common.LazyValue;
import us.minevict.mvutil.common.acf.AcfCooldowns;
import us.minevict.mvutil.common.channel.IPacketChannel;

/**
 * The base plugin for Spigot plugins using MV-Util.
 *
 * @since 3.4.0
 */
@SuppressWarnings("RedundantThrows") // They exist to show what is allowed to be thrown.
public abstract class MvPlugin
    extends JavaPlugin
    implements IMvPlugin<PaperCommandManager, Listener, MinevictusUtilsSpigot> {
  private final LazyValue<PaperCommandManager> acf = new LazyValue<>(this::constructAcf);
  private final List<BukkitTask> tasks = new ArrayList<>();
  private final List<IPacketChannel<?, ?>> channels = new ArrayList<>();
  private PluginErrorState errorState = null;
  private TaskChainFactory taskChainFactory = null;
  private String databaseName;
  private final LazyValue<Database> database = new LazyValue<>(this::constructDatabase);

  @Override
  public final void onLoad() {
    databaseName = getName().toLowerCase();

    try {
      if (!load()) {
        errorState = PluginErrorState.LOAD;
        getLogger().warning("Disabling plugin...");
        setEnabled(false);
      }
    } catch (Exception ex) {
      errorState = PluginErrorState.LOAD;
      getLogger().severe("Encountered exception upon loading:");
      ex.printStackTrace();
      getLogger().warning("Disabling plugin...");
      setEnabled(false);
    }

    if (errorState != null) {
      shutdownSafely(false);
    }
  }

  @Override
  public final void onEnable() {
    if (errorState != null) {
      return;
    }

    taskChainFactory = BukkitTaskChainFactory.create(this);

    try {
      if (!enable()) {
        errorState = PluginErrorState.ENABLE;
        getLogger().warning("Disabling plugin...");
        setEnabled(false);
      }
    } catch (Exception ex) {
      errorState = PluginErrorState.ENABLE;
      getLogger().severe("Encountered exception upon enabling:");
      ex.printStackTrace();
      getLogger().warning("Disabling plugin...");
      setEnabled(false);
    }

    if (!isEnabled()) {
      shutdownSafely(true);
    }
  }

  @Override
  public final void onDisable() {
    if (errorState != null) {
      return;
    }

    shutdownSafely(false);

    try {
      disable();
    } catch (Exception ex) {
      getLogger().severe("Encountered exception upon disabling:");
      ex.printStackTrace();
    }

    taskChainFactory.shutdown(5, TimeUnit.SECONDS);
  }

  private void shutdownSafely(boolean doChain) {
    if (doChain) {
      taskChainFactory.shutdown(5, TimeUnit.SECONDS);
    }
    tasks.forEach(BukkitTask::cancel);
    HandlerList.unregisterAll(this);
    acf.getIfInitialised().ifPresent(PaperCommandManager::unregisterCommands);
    channels.forEach(IPacketChannel::unregisterIncoming);
  }

  /**
   * {@inheritDoc}
   */
  @Override
  @NotNull
  public MinevictusUtilsSpigot getMvUtil() {
    return MinevictusUtilsSpigot.getInstance();
  }

  @NotNull
  private PaperCommandManager constructAcf() {
    return getMvUtil().prepareAcf(new PaperCommandManager(this));
  }

  @NotNull
  private Database constructDatabase() {
    return getMvUtil().prepareDatabase(this.databaseName, this);
  }

  /**
   * Gets a {@link PaperCommandManager} linked to this plugin.
   * <p>
   * This has already been {@link MinevictusUtilsSpigot#prepareAcf prepared} and is only constructed once gotten.
   *
   * @return A newly constructed or cached {@link PaperCommandManager} for this plugin.
   */
  @Override
  @NotNull
  public final PaperCommandManager getAcf() {
    return acf.getValue();
  }

  /**
   * {@inheritDoc}
   */
  @Override
  @NotNull
  public Database getDatabase() {
    return database.getValue();
  }

  /**
   * {@inheritDoc}
   */
  @Override
  @NotNull
  public String getDatabaseName() {
    return databaseName;
  }

  /**
   * {@inheritDoc}
   */
  @Override
  public void setDatabaseName(String databaseName) {
    this.databaseName = databaseName;
  }

  /**
   * Registers commands to this plugin's {@link PaperCommandManager}.
   * <p>
   * This overwrites other commands in the same names by default.
   *
   * @param commands The commands to register.
   */
  @Override
  public final void registerCommands(@NotNull BaseCommand... commands) {
    registerCommands(true, commands);
  }

  /**
   * Registers commands to this plugin's {@link PaperCommandManager}.
   *
   * @param force    Whether to overwrite other commands in the same names.
   * @param commands The commands to register.
   */
  public final void registerCommands(boolean force, @NotNull BaseCommand... commands) {
    var acf = getAcf();
    for (var command : commands) {
      acf.registerCommand(command, force);
    }
  }

  /**
   * {@inheritDoc}
   */
  @Override
  public void listeners(@NotNull Listener... listeners) {
    var pluginManager = getServer().getPluginManager();
    for (var listener : listeners) {
      pluginManager.registerEvents(listener, this);
    }
  }

  /**
   * {@inheritDoc}
   */
  @Override
  @NotNull
  public String getVersion() {
    return getDescription().getVersion();
  }

  /**
   * Add the given tasks to a cancelling routine upon disable.
   *
   * @param tasks The tasks to cancel.
   */
  public final void tasks(@NotNull BukkitTask... tasks) {
    this.tasks.addAll(Arrays.asList(tasks));
  }

  /**
   * Creates a new task chain for this plugin.
   *
   * @param <T> The type of the chain.
   * @return A newly constructed task chain.
   */
  public <T> TaskChain<T> newTaskChain() {
    return taskChainFactory.newChain();
  }

  /**
   * Creates a new task chain for this plugin.
   *
   * @param name The shared name for this chain.
   * @param <T>  The type of the chain.
   * @return A newly constructed task chain.
   */
  public <T> TaskChain<T> newTaskChain(@NotNull String name) {
    return taskChainFactory.newSharedChain(name);
  }

  /**
   * Set up the tables for these cooldowns.
   *
   * @param database  The database to set up the tables within.
   * @param cooldowns The cooldowns to setup tables for.
   * @since 3.6.0
   */
  public void setupCooldowns(@NotNull Database database, @NotNull AcfCooldowns[] cooldowns) {
    for (var cooldown : cooldowns) {
      cooldown.setupTable(database).exceptionally(ex -> {
        getLogger().warning("Exception when setting up cooldown");
        ex.printStackTrace();
        return null;
      });
    }
  }

  /**
   * {@inheritDoc}
   */
  @Override
  @NotNull
  public <P, C extends IPacketChannel<P, ?>> C packetChannel(@NotNull C channel) {
    channels.add(channel);
    return channel;
  }
}

File: spigot/src/main/kotlin/us/minevict/mvutil/spigot/MvPlugin.kt
/**
 * MV-Util
 * Copyright (C) 2020 Mariell Hoversholm, Nahuel Dolores
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package us.minevict.mvutil.spigot

import co.aikar.commands.BaseCommand
import co.aikar.commands.PaperCommandManager
import com.google.common.reflect.TypeToken
import org.bukkit.event.HandlerList
import org.bukkit.event.Listener
import org.bukkit.plugin.java.JavaPlugin
import org.bukkit.scheduler.BukkitTask
import us.minevict.mvutil.common.IMvPlugin
import us.minevict.mvutil.common.channel.IPacketChannel
import us.minevict.mvutil.common.config.TomlConfiguration
import us.minevict.mvutil.common.config.TomlDelegatedConfigurationValue
import us.minevict.mvutil.common.ext.typeToken
import us.minevict.mvutil.spigot.permissions.PermissionsDSL
import java.io.File
import java.io.InputStream
import java.util.logging.Logger
import kotlin.reflect.KClass

/**
 * The base plugin for Spigot plugins using MV-Util.
 *
 * @since 5.0.0
 */
abstract class MvPlugin : JavaPlugin(),
    IMvPlugin<PaperCommandManager, Listener, MinevictusUtilsSpigot> {
    override val platformLogger: Logger
        get() = logger

    private val tomlConfigDelegate: TomlConfiguration<MvPlugin> by lazy {
        TomlConfiguration(this, File(dataFolder, "config.toml"), "config.toml")
    }
    override val tomlConfig: TomlConfiguration<*>
        get() = tomlConfigDelegate

    private val acfDelegate = lazy {
        mvUtil.prepareAcf(PaperCommandManager(this))
    }
    override val acf by acfDelegate

    private val databaseDelegate = lazy {
        mvUtil.prepareDatabase(databaseName, this)
    }
    override val database by databaseDelegate

    override val mvUtil: MinevictusUtilsSpigot
        get() = MinevictusUtilsSpigot.instance

    override lateinit var databaseName: String

    private val tasks = mutableListOf<BukkitTask>()
    private val channels = mutableListOf<IPacketChannel<*, *>>()
    private var errorState: IMvPlugin.PluginErrorState? = null

    override val pluginName: String
        get() = description.name

    override val version: String
        get() = description.version

    final override fun onLoad() {
        databaseName = pluginName.toLowerCase().replace('-', '_')

        try {
            if (!load()) {
                errorState = IMvPlugin.PluginErrorState.LOAD
                logger.warning("Disabling plugin...")
                isEnabled = false
            }
        } catch (ex: Exception) {
            errorState = IMvPlugin.PluginErrorState.LOAD
            logger.severe("Encountered exception upon loading:")
            ex.printStackTrace()
            logger.warning("Disabling plugin...")
            isEnabled = false
        }

        if (errorState != null) {
            shutdownSafely()
        }
    }

    final override fun onEnable() {
        if (errorState != null) return

        try {
            if (!enable()) {
                errorState = IMvPlugin.PluginErrorState.ENABLE
                logger.warning("Disabling plugin...")
                isEnabled = false
            }
        } catch (ex: Exception) {
            errorState = IMvPlugin.PluginErrorState.ENABLE
            logger.severe("Encountered exception upon enabling:")
            ex.printStackTrace()
            logger.warning("Disabling plugin...")
            isEnabled = false
        }

        if (!isEnabled) {
            shutdownSafely()
        }
    }

    final override fun onDisable() {
        if (errorState != null) return

        try {
            disable()
        } catch (ex: Exception) {
            logger.severe("Encountered exception upon disabling:")
            ex.printStackTrace()
        }

        shutdownSafely()
    }

    private fun shutdownSafely() {
        tasks.forEach(BukkitTask::cancel)
        HandlerList.unregisterAll(this)
        if (acfDelegate.isInitialized())
            acf.unregisterCommands()
        channels.forEach(IPacketChannel<*, *>::unregisterIncoming)
        if (databaseDelegate.isInitialized())
            database.close()
    }

    override fun listeners(vararg listeners: Listener) {
        listeners.forEach {
            server.pluginManager.registerEvents(it, this)
        }
    }

    override fun <P, C : IPacketChannel<P, *>?> packetChannel(channel: C): C {
        channels.add(channel as IPacketChannel<*, *>)
        return channel
    }

    fun tasks(vararg tasks: BukkitTask) {
        this.tasks.addAll(tasks)
    }

    override fun registerCommands(vararg commands: BaseCommand) {
        registerCommands(true, *commands)
    }

    fun registerCommands(force: Boolean, vararg commands: BaseCommand) {
        commands.forEach {
            acf.registerCommand(it, force)
        }
    }

    fun permissions(block: PermissionsDSL.() -> Unit) {
        PermissionsDSL().also(block)
    }

    override fun getPluginResourceAsInputStream(resourceName: String): InputStream? = getResource(resourceName)

    /**
     * Delegate the config value requested with type [R] to the name or the property name with an optional default.
     *
     * @return The config value delegate.
     * @since 5.2.0
     */
    @Suppress("DEPRECATION")
    inline fun <reified R : Any> config(
        name: String? = null,
        noinline default: () -> R? = { null }
    ) = __config_internal_getter(name, typeToken<R>(), R::class, default)

    /**
     * @since 5.2.0
     */
    @Suppress("UnstableApiUsage", "DeprecatedCallableAddReplaceWith")
    @Deprecated("This is an internal API.")
    private fun <R : Any> config(
        name: String?,
        typeToken: TypeToken<out R>,
        type: KClass<out R>,
        default: () -> R?
    ) = TomlDelegatedConfigurationValue(name, typeToken, type, default)

    /**
     * @since 5.2.0
     */
    @Suppress("FunctionName", "DEPRECATION", "DeprecatedCallableAddReplaceWith", "UnstableApiUsage")
    @PublishedApi
    @Deprecated("This is an internal API.")
    internal fun <R : Any> __config_internal_getter(
        name: String?,
        typeToken: TypeToken<out R>,
        type: KClass<out R>,
        default: () -> R?
    ) = config(name, typeToken, type, default)
}